### PR TITLE
Revert "Revert "Deflakey test advanced 9""

### PR DESCRIPTION
**Pull request by fishbone**
Reverts ray-project/ray#35090

async_wait doesn't work in the same way in linux/mac/windows. Using another method to monitor the connection.

Only updates are in the [commit](https://github.com/ray-project/ray/pull/35091/files/972cbf9bfdbcef7f18dab9c6b8cf1e411f5b7e94..e84f6f5ad5a016b501c8fdf68f5e5ce7dbbf1812)

The new method listens to two events:
- EOF
- Connection failure.

And this should cover all platforms

### CODE REVIEW

**Comment by iycheng** [comment]
Hmmm. linux not working in this way.. :(

**Review by rkooo567 — COMMENTED** [review]
Have some questions/concerns regarding reading sockets from this new method + original client. 

Alternatively, what about we just handle the failure here? 

https://github.com/ray-project/ray/blob/c70331d95c249ae79ee708b97095780f63912f5e/src/ray/common/client_connection.cc#L515

So if it is failed & call DisconnectClient, we call Close() inside client_connection.cc and invoke a callback (which reports the worker failure). We can also make the additional boolean like "is_disconnected" to make this idempotent?

**Review by rkooo567 — COMMENTED** [review]
Have some questions/concerns regarding reading sockets from this new method + original client. 

Alternatively, what about we just handle the failure here? 

https://github.com/ray-project/ray/blob/c70331d95c249ae79ee708b97095780f63912f5e/src/ray/common/client_connection.cc#L515

So if it is failed & call DisconnectClient, we call Close() inside client_connection.cc and invoke a callback (which reports the worker failure). We can also make the additional boolean like "is_disconnected" to make this idempotent?

File: src/ray/common/client_connection.h [hunk 1]
```diff
@@ -125,6 +126,8 @@ class ServerConnection : public std::enable_shared_from_this<ServerConnection> {
 
   std::string DebugString() const;
 
+  void AsyncWaitTerminated(std::function<void()> callback);
```
[the remark this answers is not in the document]
Comment (iycheng) [reply src/ray/common/client_connection.h:129]: Actually, we are not closing it actively I believe. It's waiting until bad thing happened passivately. Maybe we shouldn't include close there?

**Comment by iycheng** [comment]
> Have some questions/concerns regarding reading sockets from this new method + original client.
> 
> Alternatively, what about we just handle the failure here?
> 
> https://github.com/ray-project/ray/blob/c70331d95c249ae79ee708b97095780f63912f5e/src/ray/common/client_connection.cc#L515
> 
> So if it is failed & call DisconnectClient, we call Close() inside client_connection.cc and invoke a callback (which reports the worker failure). We can also make the additional boolean like "is_disconnected" to make this idempotent?

The issue here is that, once we received DisconnectClient, the draining part is not handled any more. And the draining is similar as this one.

Maybe I can change that logic there and make the change simpler and easier to understand. Let me give it a try.

File: src/ray/common/client_connection.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rkooo567) [reply src/ray/common/client_connection.h:129]: I see. I thought we close it because we have Close() method inside this Wait method (after async_wait).

**Comment by rkooo567** [comment]
Sgtm! I scheduled a short meeting tomorrow to talk more about "he draining part is not handled any more. And the draining is similar as this one.

"

**Comment by iycheng** [comment]
Offline synced. I'm going to close this one and open two things:
- refactoring process by using boost process
- make a short term fix by exchanging the lines in core worker.

**Pull request closed without merging**